### DISCUSSION TITLE
How to retry failing tests?

### DISCUSSION THREAD

**Discussion by mmateevadeltek** [xunit/xunit#2905] [Question] [answered]
Hi, I have implemented a custom xunit test framework, inheriting from xunit classes and I am executing in parallel. For some reasons (using 3rd party libraries, that do not work well in parallel), there are few tests that fail on the first time they are executed, but if they are run again, they pass. My idea is to retry failing tests for example 3 times. This is the code that I used for that idea:

```
internal class ITTestMethodRunner : XunitTestMethodRunner
{
	private const int numberOfRetries = 3;
	private readonly IMessageSink diagnosticMessageSink;

	public ITTestMethodRunner(ITestMethod testMethod,
		IReflectionTypeInfo @class,
		IReflectionMethodInfo method,
		IEnumerable<IXunitTestCase> testCases,
		IMessageSink diagnosticMessageSink,
		IMessageBus messageBus,
		ExceptionAggregator aggregator,
		CancellationTokenSource cancellationTokenSource,
		object[] constructorArguments)
		: base(testMethod,
			  @class,
			  method,
			  testCases,
			  diagnosticMessageSink,
			  messageBus,
			  aggregator,
			  cancellationTokenSource,
			  constructorArguments)
	{
		this.diagnosticMessageSink = diagnosticMessageSink;
	}

	protected override async Task<RunSummary> RunTestCaseAsync(IXunitTestCase testCase)
	{
		var summary = new RunSummary();			
		var currentRetry = 0;

		while (currentRetry <= numberOfRetries)
		{
			try
			{
				var tempSummary = await base.RunTestCaseAsync(testCase);

				// If test passed, we can break the loop and return the summary
				if (tempSummary.Failed == 0)
				{
					summary = tempSummary;
					break;
				}
				// If it failed but we have retries left, we prepare for a retry
				else if (currentRetry < numberOfRetries)
				{
					diagnosticMessageSink.OnMessage(new DiagnosticMessage($"Retrying test {testCase.DisplayName} - attempt {currentRetry + 1} of {numberOfRetries}"));
				}
			}
			catch (Exception ex)
			{
				if (currentRetry >= numberOfRetries)
				{
					// If we've exhausted retries, we need to let the exception bubble up
					throw;
				}
				// Log the exception details if you wish to
			}
			currentRetry++;
		}

		return summary;
	}
}
```
I have a test method that is always failing on the first run, and succeeding thereafter, but when I run it, I get this summary:

`========== Test run finished: 2 Tests (1 Passed, 1 Failed, 0 Skipped) run in 10.6 sec ==========`

And the output of the test is this:

```
Test has multiple result outcomes
   1 Passed
   1 Failed

Results

    1)   Test 
      Duration: 162 ms

      Message: 
This test is designed to fail on the first run and pass thereafter.

      Stack Trace: 
IT_AccessTag.Test() line 30

      Standard Output: 
Database Db_1, Created
Run Mode: Default



    2)   Test 
      Duration: < 1 ms

      Standard Output: 
Database:  Db_1, Created
Run Mode: Default
```

Can you give me an advice how to do it properly? Thank you!

**Comment by bradwilson** [comment, marked as answer]
We have a sample that does this: https://github.com/xunit/samples.xunit/tree/main/RetryFactExample

What's happening here is that the message listener in the runner is seeing you run the test multiple times, because you're letting it see the result messages from both the failures and the successes. Our sample shows using a "delayed message bus" so that it hangs onto all the messages while the test runs, and will only ever send those messages along if the test passes, or if the fail count exceeds the limit.

**Reply by mmateevadeltek** [reply]
Thank you, @bradwilson !

Yes, I could get it working for a my custom Fact attribute, but I also use Theory, and I am having trouble to get this working. It seems I cannot use the same approach as I do for facts, because it need other parameters, right?

This is the error I get:
`System.InvalidOperationException : The test method expected 1 parameter value, but 0 parameter values were provided.`

And the code that I use is from your example:
```
public override async Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink,
												IMessageBus messageBus,
												object[] constructorArguments,
												ExceptionAggregator aggregator,
												CancellationTokenSource cancellationTokenSource)
{
	var runCount = 0;

	while (true)
	{
		// This is really the only tricky bit: we need to capture and delay messages (since those will
		// contain run status) until we know we've decided to accept the final result;
		var delayedMessageBus = new DelayedMessageBus(messageBus);

		var summary = await base.RunAsync(diagnosticMessageSink, delayedMessageBus, constructorArguments, aggregator, cancellationTokenSource);
		
		if (aggregator.HasExceptions || summary.Failed == 0 || ++runCount >= maxRetries)
		{
			delayedMessageBus.Dispose();  // Sends all the delayed messages
			return summary;
		}

		diagnosticMessageSink.OnMessage(new DiagnosticMessage("Execution of '{0}' failed (attempt #{1}), retrying...", DisplayName, runCount));
	}
}
```

**Comment by mmateevadeltek** [comment]
Hi, it is me again, I was able to do that for Theory as well.

The first thing is to change the constructor of the **RetryTestCase** to pass **testMethodArguments** as well, like this:
```
public RetryTestCase(int maxRetries,
	IMessageSink diagnosticMessageSink,
	TestMethodDisplay testMethodDisplay,
	TestMethodDisplayOptions defaultMethodDisplayOptions,
	ITestMethod testMethod,
	object[] testMethodArguments = null)
	: base(diagnosticMessageSink,
		  testMethodDisplay,
		  defaultMethodDisplayOptions,
		  testMethod,
		  testMethodArguments)
{
	this.maxRetries = maxRetries;
}
```

Second, I created my own **TheoryDiscoverer** in which I am overriding **CreateTestCasesForDataRow** where I can pass the arguments:

```
public class ITTheoryDiscoverer : TheoryDiscoverer
{
	public ITTheoryDiscoverer(IMessageSink diagnosticMessageSink)
		: base(diagnosticMessageSink)
	{ }

	protected override IEnumerable<IXunitTestCase> CreateTestCasesForDataRow(ITestFrameworkDiscoveryOptions discoveryOptions, 
		ITestMethod testMethod,
		IAttributeInfo theoryAttribute,
		object[] dataRow)
	{
		var testCases = base.CreateTestCasesForDataRow(discoveryOptions, testMethod, theoryAttribute, dataRow);

		foreach(var testCase in testCases)
		{
			var maxRetries = theoryAttribute.GetNamedArgument<int>("MaxRetries");

			if (maxRetries < 1)
				maxRetries = 3;

			yield return new RetryTestCase(maxRetries,
				DiagnosticMessageSink,
				discoveryOptions.MethodDisplayOrDefault(),
				discoveryOptions.MethodDisplayOptionsOrDefault(),
				testCase.TestMethod,
				testCase.TestMethodArguments);
		}
	}
}
```

**Reply by bradwilson** [reply]
I just updated the sample to include a theory version. The changes are fairly minimal. https://github.com/xunit/samples.xunit/commit/8ad214a7f22eb50d29c45e8e4bc44234bcf6bf71

You don't need to call `base.CreateTestCasesForDataRow` since you're going to be making your own, and the data you're extracting from them, you already have access to.

**Comment by bdovaz** [comment]
@bradwilson why not even in v3 is it a built-in feature? https://github.com/xunit/samples.xunit/tree/main/v3/RetryFactExample

Both NUnit and MSTest include them:

https://docs.nunit.org/articles/nunit/writing-tests/attributes/retry.html

https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-writing-tests-attributes#retryattribute

**Reply by bradwilson** [reply]
Because, as an opinionated framework designed for unit testing, retry is not a necessary feature. It's available as a sample to illustrate how to extend xUnit.net for other types of testing scenarios.

**Reply by bdovaz** [reply]
@bradwilson what I mean is that even if it is like another NuGet package but it is something that would make sense to keep in the project itself and not as an example....

It's a bit absurd that packages like this have to exist for that reason: https://github.com/JoshKeegan/xRetry/tree/master/src/xRetry

**Reply by bradwilson** [reply]
You're drawing an arbitrary line. Should _any_ third party packages exist, or should be make them all extension packages under our control? The reality is that we have VERY limited time to work on this project, and doing this would mean not doing something else. In our judgment this is _exactly_ the kind of thing someone else should do, your judgments of absurdity notwithstanding (and not relevant).

**Reply by bdovaz** [reply]
> You're drawing an arbitrary line. Should _any_ third party packages exist, or should be make them all extension packages under our control?

I was referring to that particular case (retry attribute) because as I said, compared to the other 2 testing frameworks, it is the only one that does not have that functionality that is considered relatively basic when you do integration tests that do not have to be deterministic by having other third-party dependencies or network conditions.

> The reality is that we have VERY limited time to work on this project, and doing this would mean not doing something else. In our judgment this is _exactly_ the kind of thing someone else should do, your judgments of absurdity notwithstanding (and not relevant).

`dotnet/runtime` / `dotnet/extensions` / `dotnet/aspnetcore` use this framework with the relevance they have and you make that comment...

It is quite worrying that this project, with the use it has, depends exclusively on you and you have so little dedication. I suppose you have asked the community for help? I see your commits and the second contributor commit count (on top of that, they are from 2016) and it is a cause for concern:

https://github.com/xunit/xunit/graphs/contributors

I maintain or contribute to other projects and the same thing happens to me, but it has nothing to do with the level of use of this project compared to the ones I maintain or contribute to, of course.

**Reply by bradwilson** [reply]
> I was referring to that particular case (retry attribute) because as I said, compared to the other 2 testing frameworks, it is the only one that does not have that functionality that is considered relatively basic when you do integration tests

And as I previously mentioned, we are _difference_ precisely because we opinionated on being designed for _unit testing_ and not for _integration testing_. We enable third parties to light up integration testing scenarios rather than directly support them ourselves, and that's a reason why we have so many extensibility points, as illustrated by the Retry sample.

We don't do things just because everybody else does them; if everybody did that, then what's the point of separate frameworks?

> `dotnet/runtime` / `dotnet/extensions` / `dotnet/aspnetcore` use this framework with the relevance they have and you make that comment...

Would you prefer that I lied to you and pretended that this was some 100 person company? It's just me, in my free time. If that is concerning to you, then you should probably use MSTest, because it's the only corporate test framework available. Everything else is being done by very few people in their spare time.

> you have so little dedication

Thanks for your vote of confidence, despite the evidence you linked about how much I contribute to this project.

This is probably a good time to lock this discussion now.
